Automattic/wp-calypso
# Receipts: Convert receipt assembler types only to TypeScript

#### Proposed Changes

This is extracted from part of #68857, which was reverted in https://github.com/Automattic/wp-calypso/pull/69035. This part of the original PR only moves the receipt assembler types from JSDoc to a real TypeScript types file without making functional changes to the code. Doing this first will make it easier to see the changes to the assembler file [when trying again to convert it to TypeScript](https://github.com/Automattic/wp-calypso/pull/69107).

Because I think that the JSDoc types may not have been liberal enough, this also updates the types for `RawReceiptData.purchases` and `RawReceiptData.failed_purchases` to include `undefined` and `false`. This should offer additional protection against API endpoint data which is not quite what we expect.

#### Testing Instructions

This should change no actual code at all so it should be safe. Verify that there are no functional changes in this PR.

Verify that the [bug reported here](https://github.com/Automattic/wp-calypso/pull/69035#issuecomment-1277842949) is not present on this branch.
